torchtune/dev/rl/workers/datacollectors/sync.py

```diff
@@ -134,60 +135,92 @@ def remote_weight_updater(self, value: RemoteWeightUpdaterBase | None):
+
+        trajectories = defaultdict(list)
+        for i in range(batch_size):
+            # extract data
+            prompt_tokens = prompt_tokens_padded[i]
+            response_tokens = response_tokens_padded[i]
+            logprobs = logprobs_padded[i]
+            answer = answers[i]
+            
+            # remove padding tokens
+            prompt_mask = prompt_tokens != self._tokenizer.pad_id
+            response_mask = response_tokens != self._tokenizer.pad_id
+            prompt_len = prompt_mask.sum()
+            response_len = response_mask.sum()
+            
+            prompt_tokens = prompt_tokens[prompt_mask]
+            response_tokens = response_tokens[response_mask]
+            logprobs = logprobs[response_mask]
+            
+            # group id is used to calculate advantages
+            group_idx = i // self.cfg.grpo_samples
+            group_id = f"{self.worker_id}_{self._sequence_counter}_{group_idx}"
+            
+            # create trajectory
+            traj = UnscoredTrajectory(
+                prompt_tokens=prompt_tokens,
+                response_tokens=response_tokens,
+                prompt_len=prompt_len,
+                response_len=response_len,
+                logprobs=logprobs,
+                answer=answer,
+                sequence_id=sequence_ids[i],
+                policy_version=policy_version,
+                group_id=group_id
             )
-        else:
-            policy_version = 0
-
-        response_padding_masks = torch.eq(response_tokens, self._tokenizer.pad_id)
-        seq_lens = training.get_unmasked_sequence_lengths(response_padding_masks)
-        del response_padding_masks
-
-        # Generate unique sequence IDs for the batch
-        # FIXME: it outputs a List[List[str]] when sampling from replay buffer, with shape num_samples X 16.
-        # It should have shape num_samplesX1, so we can log a single sequence_id per sequence.
-        batch_size = query_responses.shape[0]
-        sequence_ids = NonTensorStack(
-            *[
-                f"worker{self.worker_id}_{self._sequence_counter + i}"
-                for i in range(batch_size)
-            ]
-        )
-        self._sequence_counter += batch_size
 
-        postprocessed_results = Trajectory(
-            query_responses=query_responses,
-            responses=response_tokens,
-            logprobs=logprobs,
-            ref_logprobs=None,
-            query_response_padding_masks=query_response_padding_masks,
-            seq_lens=seq_lens,
-            answers=answers,
-            policy_version=policy_version,
-            rewards=None,
-            advantages=None,
-            successes=None,
-            reward_metadata=None,
-            sequence_ids=sequence_ids,
-        )
+            # group trajectories by group_id
+            trajectories[traj.group_id].append(traj)
+
+        
+        # create GroupedUnscoredTrajectories for each group
+        grouped_trajectories = []
```

Comment:
Eventually this will need to be enabled if Monte Carlo-based advantage estimation is turned on, otherwise it will be off. EG in "normal" PPO it's off